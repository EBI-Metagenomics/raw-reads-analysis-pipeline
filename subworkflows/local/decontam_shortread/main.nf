include { BWAMEM2_MEM as BWAMEM2_ALIGN_PHIX } from '../../../modules/nf-core/bwamem2/mem/main'
include { BWAMEM2_MEM as BWAMEM2_ALIGN_HOST } from '../../../modules/nf-core/bwamem2/mem/main'
include { DECONTAMBAM as DECONTAMBAM_PHIX } from '../../../modules/local/decontambam/main'
include { DECONTAMBAM as DECONTAMBAM_HOST } from '../../../modules/local/decontambam/main'
include { COMBINEBAM as COMBINEBAM_PHIX } from '../../../modules/local/combinebam/main'
include { COMBINEBAM as COMBINEBAM_HOST } from '../../../modules/local/combinebam/main'
// include { SEQKIT_SPLIT2 as SEQKIT_SPLIT2_PHIX } from '../../../modules/nf-core/seqkit/split2/main'
// include { SEQKIT_SPLIT2 as SEQKIT_SPLIT2_HOST } from '../../../modules/nf-core/seqkit/split2/main'
include { CHUNKFASTX as CHUNKFASTX_PHIX } from '../../../modules/local/chunkfastx/main'
include { CHUNKFASTX as CHUNKFASTX_HOST } from '../../../modules/local/chunkfastx/main'

workflow DECONTAM_SHORTREAD {
    take:
    reads  // [ val(meta), path(reads) ]
    host_genome  // [ val(meta2), path(reference_genome_index_root) ]
    phix_genome  // [ val(meta3), path(phix_index_root) ]

    main:
    def ch_versions = Channel.empty()

    phix_genome_index = phix_genome
        .map{ meta, fp ->
              [meta, files("${fp}/${meta.base_dir}/${meta.files.bwa_index_prefix}.*")] }
        .first()
    phix_genome_fasta = phix_genome
        .map{ meta, fp ->
              [meta, file("${fp}/${meta.base_dir}/${meta.files.genome}")] }
        .first()
    host_genome_index = host_genome
        .map{ meta, fp ->
              [meta, files("${fp}/${meta.base_dir}/${meta.files.bwa_index_prefix}.*")] }
        .first()
    host_genome_fasta = host_genome
        .map{ meta, fp ->
              [meta, file("${fp}/${meta.base_dir}/${meta.files.genome}")] }
        .first()

    // phix_genome_index.view{ "phix_genome_index - ${it}" }
    // phix_genome_fasta.view{ "phix_genome_fasta - ${it}" }
    // reference_genome_index.view{ "reference_genome_index - ${it}" }
    // reference_genome_fasta.view{ "reference_genome_fasta - ${it}" }

    if (params.remove_phix) {

        // chunked_reads = reads
        //     .flatMap{ meta, fastas ->
        //         def chunks = []
        //         if(meta.single_end) {
        //             chunks = fastas.splitFasta(file: true, compress: true, size: 500.KB)
        //         } else {
        //             chunks = [
        //                 fastas[0].splitFasta(file: true, compress: true, size: 500.KB),
        //                 fastas[1].splitFasta(file: true, compress: true, size: 500.KB)
        //             ].transpose()
        //         }
        //         chunks.collect{ chunk -> tuple(groupKey(meta, chunks.size()), chunk) }
        //     }
        // chunked_reads.view{ "chunked_reads - ${it}" }

        // SEQKIT_SPLIT2_PHIX(reads)
        CHUNKFASTX_PHIX(reads, file("${projectDir}/bin/chunk_fastx.py"))
        chunked_reads = CHUNKFASTX_PHIX.out.reads
            .flatMap{ meta, chunks ->
                if (chunks instanceof List) {
                    def grouped_chunks = [:]
                    chunks.collect{
                        chunk ->
                        def part = (chunk.name =~ /\.(chunk-\d+)\./)[0][1]
                        if (grouped_chunks.containsKey(part)) {
                            grouped_chunks[part].add(chunk)
                        } else {
                            grouped_chunks[part] = [chunk]
                        }
                    }
                    return grouped_chunks.collect{ _part, chunk -> tuple(groupKey(meta, grouped_chunks.size()), tuple(*chunk)) }
                } else {
                    return [tuple(groupKey(meta, 1), chunks)]
                }
            }
        // chunked_reads.view{ "chunked_reads - ${it}" }

        BWAMEM2_ALIGN_PHIX(
            chunked_reads,
            phix_genome_index,
            phix_genome_fasta,
            false
        )
        ch_versions = ch_versions.mix(BWAMEM2_ALIGN_PHIX.out.versions)

        COMBINEBAM_PHIX(
            BWAMEM2_ALIGN_PHIX.out.bam.groupTuple()
        )

        DECONTAMBAM_PHIX(
            COMBINEBAM_PHIX.out.concatenated_result.map{ meta, bam ->
                [meta, bam, meta.single_end==false, "short_read_phix"]
            }
        )
        ch_versions = ch_versions.mix(DECONTAMBAM_PHIX.out.versions)


        decontaminated_reads = DECONTAMBAM_PHIX.out.unmapped_reads
        phix_stats = DECONTAMBAM_PHIX.out.stats
    }
    else {
        decontaminated_reads = reads
        phix_stats = Channel.empty()
    }

    decontaminated_reads = decontaminated_reads.map { meta, reads ->
        [meta + ['decontam_phix_read_count': (meta.single_end ? reads : reads[0]).countFastq()], reads]
    }
    // decontaminated_reads.view{ meta, _reads -> "decontaminated_phix_reads - [${meta.id}, ${meta.platform}, ${meta.single_end}] - ${meta.decontam_phix_read_count}" }
    decontaminated_reads = decontaminated_reads.filter { meta, _reads ->
        meta.decontam_phix_read_count > 0
    }

    if (host_genome != null) {

        // chunked_decontaminated_reads = decontaminated_reads
        //     .flatMap{ meta, fastas ->
        //         def chunks = []
        //         if(meta.single_end) {
        //             chunks = fastas.splitFasta(file: true, compress: true, size: 500.KB)
        //         } else {
        //             chunks = [
        //                 fastas[0].splitFasta(file: true, compress: true, size: 500.KB),
        //                 fastas[1].splitFasta(file: true, compress: true, size: 500.KB)
        //             ].transpose()
        //         }
        //         chunks.collect{ chunk -> tuple(groupKey(meta, chunks.size()), chunk) }
        //     }
        // chunked_decontaminated_reads.view{ "chunked_decontaminated_reads - ${it}" }

        // SEQKIT_SPLIT2_HOST(decontaminated_reads)
        CHUNKFASTX_HOST(decontaminated_reads, file("${projectDir}/bin/chunk_fastx.py"))
        chunked_decontaminated_reads = CHUNKFASTX_HOST.out.reads
            .flatMap{ meta, chunks ->
                if (chunks instanceof List) {
                    def grouped_chunks = [:]
                    chunks.collect{
                        chunk ->
                        def part = (chunk.name =~ /\.(chunk-\d+)\./)[0][1]
                        if (grouped_chunks.containsKey(part)) {
                            grouped_chunks[part].add(chunk)
                        } else {
                            grouped_chunks[part] = [chunk]
                        }
                    }
                    return grouped_chunks.collect{ _part, chunk -> tuple(groupKey(meta, grouped_chunks.size()), tuple(*chunk)) }
                } else {
                    return [tuple(groupKey(meta, 1), chunks)]
                }
            }
        // chunked_decontaminated_reads.view{ "chunked_decontaminated_reads - ${it}" }

        BWAMEM2_ALIGN_HOST(
            chunked_decontaminated_reads,
            host_genome_index,
            host_genome_fasta,
            false
        )
        ch_versions = ch_versions.mix(BWAMEM2_ALIGN_HOST.out.versions)

        COMBINEBAM_HOST(
            BWAMEM2_ALIGN_HOST.out.bam.groupTuple()
        )

        DECONTAMBAM_HOST(
            COMBINEBAM_HOST.out.concatenated_result.map{ meta, bam ->
                [meta, bam, meta.single_end==false, "short_read_host"]
            }
        )
        ch_versions = ch_versions.mix(DECONTAMBAM_HOST.out.versions)

        decontaminated_reads = DECONTAMBAM_HOST.out.unmapped_reads
        host_stats = DECONTAMBAM_HOST.out.stats
    }
    else {
        host_stats = Channel.empty()
    }

    decontaminated_reads = decontaminated_reads.map { meta, reads ->
        [meta + ['decontam_host_read_count': (meta.single_end ? reads : reads[0]).countFastq()], reads]
    }
    // decontaminated_reads.view{ meta, _reads -> "decontaminated_host_reads - [${meta.id}, ${meta.platform}, ${meta.single_end}] - ${meta.decontam_host_read_count}" }
    decontaminated_reads = decontaminated_reads.filter { meta, _reads ->
        meta.decontam_host_read_count > 0
    }

    emit:
    decontaminated_reads = decontaminated_reads
    host_stats           = host_stats
    phix_stats           = phix_stats
    versions             = ch_versions
}
