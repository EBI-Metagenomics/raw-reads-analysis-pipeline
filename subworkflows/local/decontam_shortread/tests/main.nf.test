nextflow_workflow {

    name "Test Subworkflow DECONTAM_SHORTREAD"
    script "../main.nf"
    workflow "DECONTAM_SHORTREAD"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/decontam_shortread"
    tag "bwamem"
    tag "bwamem/align"

    test("mini paired-end sample output should match snapshot") {

        when {
	    params {
	        remove_phix = true
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': false, 'platform': 'wgs'],
		    [
		        file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_1.fq.gz", checkIfExists: true),
		        file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_2.fq.gz", checkIfExists: true),
		    ]
		])
		input[1] = Channel.of([
		    ['id': 'host_genome', 'base_dir': '.', 'files': ['bwa_index_prefix': 'hg38_mini.fa', 'genome': 'hg38_mini.fa']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini", checkIfExists: true)
		])
		input[2] = Channel.of([
		    ['id': 'phix', 'base_dir': '.', 'files': ['bwa_index_prefix': 'MN385565.fasta', 'genome': 'MN385565.fasta']],
		    file("${projectDir}/tests/fixtures/dbs/phix", checkIfExists: true)
		])
                """
            }
        }

        then {
	    def reads1 = path(workflow.out.decontaminated_reads.get(0)[1][0])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

	    def reads2 = path(workflow.out.decontaminated_reads.get(0)[1][1])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

            assert workflow.success
            assert snapshot(
                reads1, reads2, 
		// workflow.out.decontaminated_reads,
	        // workflow.out.host_stats, 
		// workflow.out.phix_stats, 
		workflow.out.versions
	    ).match()
        }

    }

    test("mini single-end sample output should match snapshot") {

        when {
	    params {
	        remove_phix = false
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': true, 'platform': 'wgs'],
		    file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_1.fq.gz", checkIfExists: true),
		])
		input[1] = Channel.of([
		    ['id': 'host_genome', 'base_dir': '.', 'files': ['bwa_index_prefix': 'hg38_mini.fa', 'genome': 'hg38_mini.fa']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini", checkIfExists: true)
		])
		input[2] = Channel.of([
		    ['id': 'phix', 'base_dir': '.', 'files': ['bwa_index_prefix': 'MN385565.fasta', 'genome': 'MN385565.fasta']],
		    file("${projectDir}/tests/fixtures/dbs/phix", checkIfExists: true)
		])
                """
            }
        }

        then {
	    def reads1 = path(workflow.out.decontaminated_reads.get(0)[1])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

            assert workflow.success
            assert snapshot(
                reads1,  
		// workflow.out.decontaminated_reads,
	        // workflow.out.host_stats, 
		// workflow.out.phix_stats, 
		workflow.out.versions
	    ).match()
        }

    }

    test("mini sample with added hg38 synthetic reads output should match snapshot") {

        when {
	    params {
	        remove_phix = true
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': false, 'platform': 'wgs'],
		    [
		        file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_plus_hg38_1.fq.gz", checkIfExists: true),
		        file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_plus_hg38_2.fq.gz", checkIfExists: true),
		    ]
		])
		input[1] = Channel.of([
		    ['id': 'host_genome', 'base_dir': '.', 'files': ['bwa_index_prefix': 'hg38_mini.fa', 'genome': 'hg38_mini.fa']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini", checkIfExists: true)
		])
		input[2] = Channel.of([
		    ['id': 'phix', 'base_dir': '.', 'files': ['bwa_index_prefix': 'MN385565.fasta', 'genome': 'MN385565.fasta']],
		    file("${projectDir}/tests/fixtures/dbs/phix", checkIfExists: true)
		])
                """
            }
        }

        then {
	    def reads1 = path(workflow.out.decontaminated_reads.get(0)[1][0])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

	    def reads2 = path(workflow.out.decontaminated_reads.get(0)[1][1])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

            assert workflow.success
            assert snapshot(
                reads1, reads2, 
		// workflow.out.decontaminated_reads,
	        // workflow.out.host_stats, 
		// workflow.out.phix_stats, 
		workflow.out.versions
	    ).match()
        }

    }

    test("hg38 synthetic reads output should match snapshot and not fail") {

        when {
	    params {
	        remove_phix = false
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': false, 'platform': 'wgs'],
		    [
		        file("${projectDir}/tests/fixtures/decontam_shortread/hg38_reads_1.fq.gz", checkIfExists: true),
		        file("${projectDir}/tests/fixtures/decontam_shortread/hg38_reads_2.fq.gz", checkIfExists: true),
		    ]
		])
		input[1] = Channel.of([
		    ['id': 'host_genome', 'base_dir': '.', 'files': ['bwa_index_prefix': 'hg38_mini.fa', 'genome': 'hg38_mini.fa']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini", checkIfExists: true)
		])
		input[2] = Channel.of([
		    ['id': 'phix', 'base_dir': '.', 'files': ['bwa_index_prefix': 'MN385565.fasta', 'genome': 'MN385565.fasta']],
		    file("${projectDir}/tests/fixtures/dbs/phix", checkIfExists: true)
		])
                """
            }
        }

        then {
            assert workflow.success
	    assert workflow.out.decontaminated_reads.size()==0
            assert snapshot(
		// workflow.out.decontaminated_reads,
	        // workflow.out.host_stats, 
		// workflow.out.phix_stats, 
		workflow.out.versions
	    ).match()
        }

    }

}
