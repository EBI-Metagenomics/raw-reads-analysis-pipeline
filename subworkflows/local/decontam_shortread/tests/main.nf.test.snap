{
    "mini single-end sample output should match snapshot": {
        "content": [
            [
                "@B@BCEG>FCG>GGGG@E99CEGBGB/BEC@G@FF@@G11C1EGGA9CFGGGGCF@=FCFGGGGGEGGGGG=DGB@/FGCGBC9EEEGG/9C/EGGGGGG/8E<.E@D/6.CGGC=D/C@CE6D6",
                "@BBBBFFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGEGGGGGGGGGCFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG",
                "@ERR10889147.101689",
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.239847",
                "@ERR10889147.280640",
                "@ERR10889147.313339",
                "@ERR10889147.350555",
                "@ERR10889147.374007",
                "@ERR10889147.377961",
                "@ERR10889147.398989",
                "@ERR10889147.421340",
                "@ERR10889147.421831",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.541000",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.56958",
                "@ERR10889147.63497",
                "@ERR10889147.646304",
                "@ERR10889147.65722",
                "@ERR10889147.662821",
                "@ERR10889147.76311",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "versions.yml:md5,d82afcd54b2afe327671878ed9827d52",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T13:03:47.401674437"
    },
    "mini paired-end sample output should match snapshot": {
        "content": [
            [
                "@B@BCEG>FCG>GGGG@E99CEGBGB/BEC@G@FF@@G11C1EGGA9CFGGGGCF@=FCFGGGGGEGGGGG=DGB@/FGCGBC9EEEGG/9C/EGGGGGG/8E<.E@D/6.CGGC=D/C@CE6D6",
                "@BBBBFFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGEGGGGGGGGGCFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG",
                "@ERR10889147.101689",
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.239847",
                "@ERR10889147.280640",
                "@ERR10889147.313339",
                "@ERR10889147.350555",
                "@ERR10889147.374007",
                "@ERR10889147.377961",
                "@ERR10889147.398989",
                "@ERR10889147.421340",
                "@ERR10889147.421831",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.541000",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.56958",
                "@ERR10889147.63497",
                "@ERR10889147.646304",
                "@ERR10889147.65722",
                "@ERR10889147.662821",
                "@ERR10889147.76311",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "@B@BB>GEGGGEE1>FDGC@1=EBFGDC@:B<E@>F>@@CFGCGE:B<C@:00=FGG0C=1<EFGE::<@E>FGGEE@:FG0C1FGD9EG/CD00F@D8D0FCFDDGG@088=FDGGCG/CE@",
                "@ERR10889147.101689",
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.239847",
                "@ERR10889147.280640",
                "@ERR10889147.313339",
                "@ERR10889147.350555",
                "@ERR10889147.374007",
                "@ERR10889147.377961",
                "@ERR10889147.398989",
                "@ERR10889147.421340",
                "@ERR10889147.421831",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.541000",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.56958",
                "@ERR10889147.63497",
                "@ERR10889147.646304",
                "@ERR10889147.65722",
                "@ERR10889147.662821",
                "@ERR10889147.76311",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "versions.yml:md5,2754cc3b6e8cabeda717483392df165e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,d82afcd54b2afe327671878ed9827d52",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T13:00:37.281796561"
    },
    "mini sample with added hg38 synthetic reads output should match snapshot": {
        "content": [
            [
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.280640",
                "@ERR10889147.350555",
                "@ERR10889147.377961",
                "@ERR10889147.421340",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.662821",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.280640",
                "@ERR10889147.350555",
                "@ERR10889147.377961",
                "@ERR10889147.421340",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.662821",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "versions.yml:md5,2754cc3b6e8cabeda717483392df165e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,a50f344db10ec06ebc0ff5f4f247bf6e",
                "versions.yml:md5,d82afcd54b2afe327671878ed9827d52",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T13:06:22.346410691"
    },
    "hg38 synthetic reads output should match snapshot and not fail": {
        "content": [
            [
                "versions.yml:md5,d82afcd54b2afe327671878ed9827d52",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3",
                "versions.yml:md5,f2f76a6c8010db22d8e4a1f4e48334d3"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T13:08:08.993013794"
    }
}